### DISCUSSION TITLE
Convert from punycode to unicode

### DISCUSSION THREAD

**Discussion by bagder** [curl/trurl#228] [Ideas]
## This works already
~~~
$ trurl 'räksmörgås.se' --punycode
http://xn--rksmrgs-5wao1o.se/
~~~

## This does not work
~~~
$ trurl 'http://xn--rksmrgs-5wao1o.se/' --unpunify
http://räksmörgås/

**Comment by jacobmealey** [comment]
Would implementing this require changes to libcurl to add support for going from punycode to unicode? I have been poking around with it a bit this morning and it seems to only go in one direction - though I may very well be looking in the wrong locations. Or instead of changes to libcurl having some standalone function that implements the decode steps from RFC 3492 in trurl?

**Reply by bagder** [reply]
I was pondering a bit on this as well. I believe that there are two alternative routes:

1. Add the functionality to libcurl, which uses libidn2 (primarily) for punycode conversions already (and it has this ability). Then make trurl use that.
2. Add the functionality using a third party library (like libidn2) directly so that the feature can be offered without relying on libcurl. Will require some juggling to complicate the build procedure and will possibly duplicate functionality that is better put into libcurl instead.

Route 2 is faster to get functionality into trurl, but my gut says route 1 is probably more sensible...

**Reply by jacobmealey** [reply]
I agree that taking route 1 seems to be the the sensible option. The only downside to taking route 1 is that it would require another minimum version bump in order to be used, which is fine because we already have a few different flags being set in order to build various features based on the version of libcurl available. 

I also think keeping the build simple and only relying on one library is a good route to take.

**Reply by bagder** [reply]
The journey has begun: https://github.com/curl/curl/pull/11655

**Reply by jacobmealey** [reply]
awesome! I was going to start writing some code this weekend if no one else started. I've been slowly going through urlapi.c (and friends) to get an idea of where the changes should take place, glad to see I was looking in the right spots!

**Reply by bagder** [reply]
It has been merged into curl's git master now, which thus makes it possible to start working on a trurl implementation. Do you want to take that? It'll be a good extra check that the libcurl API for this also works as expected - before it ships.

**Reply by jacobmealey** [reply]
yeah, I can start working on that this week

**Reply by jacobmealey** [reply]
I've started a pull request here https://github.com/curl/trurl/pull/231

**Comment by bagder** [comment]
Maybe `--unpunify` is not the best name for the option...

**Reply by jacobmealey** [reply]
I've been mulling this over as well. I'm not sure if it's any better but maybe something like  `--unicode`?  I think it would make sense this way because theres `--punycode` to encode the url in punycode, so it makes sense that we would use `--unicode` to encode it in unicode?
